cmake/ExperimentalEmscripten.cmake

```diff
@@ -0,0 +1,112 @@
+
+    message(TRACE "Appending ${EMSDK_ROOT}/upstream/emscripten/cmake/Modules/Platform to CMAKE_MODULE_PATH")
+    list(APPEND CMAKE_MODULE_PATH "${EMSDK_ROOT}/upstream/emscripten/cmake/Modules/Platform")
+
+    include (Emscripten)
+else()
+    message(FATAL_ERROR "Could not find emsdk installation. Please install the Emscripten SDK.\nhttps://emscripten.org/docs/getting_started/downloads.html ")
+endif()
+
+#
+# Explanation of flags used:
+#
+# `-pthread`
+# STLab uses threads. Without this, the tests will not compile.
+#
+# `-fwasm-exceptions`:
+# STLab uses exceptions. Without this, the tests error out with:
+#
+#   Pthread 0x005141d0 sent an error! http://localhost:6931/<throwing test>: uncaught exception: 10570976 \
+#   - Exception catching is disabled, this exception cannot be caught.
+#
+# `-sSUPPORT_LONGJMP=wasm`
+# Enables experimental support for LONGJMP in functions which may throw exceptions.
+# Without this, boost would not compile (LLVM errored out). 
+#
+# `-sEXIT_RUNTIME=1`
+# Indicates the runtime environment (node) should exit when our main() returns.
+#
+# `-sINITIAL_MEMORY=300MB` 
+# Without this, the tests OOM. The first sign is an error out with the following:
+#
+#   Pthread 0x0058faf8 sent an error! http://localhost:6931/<test>: RuntimeError: unreachable executed
+#
+# If we run the problematic test in a browser with `emrun`, we can see some javascript errors that explain:
+#
+#   Aborted(Cannot enlarge memory arrays to size 17457152 bytes (OOM). Either
+#   (1) compile with -sINITIAL_MEMORY=X with X higher than the current value 16777216, 
+#   (2) compile with -sALLOW_MEMORY_GROWTH which allows increasing the size at runtime, or 
+#   (3) if you want malloc to return  NULL (0) instead of this abort, compile with -sABORTING_MALLOC=0)
+#
+# Note that (2) is not an option because pthread cannot yet be combined with -sALLOW_MEMORY_GROWTH:
+# See https://github.com/WebAssembly/design/issues/1271
+# Smaller values (150MB, 200MB) produce intermittent failures. 300MB was chosen to give enough headroom for
+# tests written in the future. 
+#
+# `-sPTHREAD_POOL_SIZE=32` 
+# Without this, the tests deadlock. Lower values were tested. 
+# 8 threads deadlocked consistently, 16 threads passed consistently. 
+# 32 was chosen to give enough headroom for tests written in the future. 
+#
+# `-sPROXY_TO_PTHREAD`
+# This flag wraps our executable's main function in a pthread. 
+# Without this, we exhaust the thread pool very quickly. The error looks like this:
+#
+#   Tried to spawn a new thread, but the thread pool is exhausted.
+#   This might result in a deadlock unless some threads eventually exit or the code explicitly breaks out to the event loop.
+# 
+# You can read more about the setting here: https://emscripten.org/docs/porting/pthreads.html#blocking-on-the-main-browser-thread
+#
+
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -pthread -fwasm-exceptions -sSUPPORT_LONGJMP=wasm")
```

Comment:
One minor suggestion would be to put the documentation right next to where the flag is being set. E.g.

```cmake
set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -pthread")
# STLab uses threads. Without this, the tests will not compile.

set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fwasm-exceptions")
# STLab uses exceptions. Without this, the tests error out with:
#
#   Pthread 0x005141d0 sent an error! http://localhost:6931/<throwing test>: uncaught exception: 10570976 \
#   - Exception catching is disabled, this exception cannot be caught.
```

Reply:
I'm happy to split them up - but my instinct would be to put the comment before the command. Does that matter to you?

Reply:
Sounds good to me.